File: flutter_client/lib/bloc/chat_bloc.dart
// ABOUTME: BLoC managing chat state and API server communication.
// ABOUTME: Consumes client event streams from API server for real-time updates.

import 'package:bloc/bloc.dart';
import 'package:logger/logger.dart';
import '../models/message.dart';
import '../models/agent_event.dart';
import '../services/api_client.dart';
import 'chat_event.dart';
import 'chat_state.dart';
import 'map_bloc.dart';
import 'map_event.dart';

class ChatBloc extends Bloc<ChatEvent, ChatState> {
  final ApiClient _apiClient;
  final MapBloc _mapBloc;
  final Logger _logger;

  ChatBloc({
    required ApiClient apiClient,
    required MapBloc mapBloc,
    Logger? logger,
  })  : _apiClient = apiClient,
        _mapBloc = mapBloc,
        _logger = logger ?? Logger(),
        super(ChatState()) {
    on<SendMessage>(_onSendMessage);
  }

  Future<void> _onSendMessage(
    SendMessage event,
    Emitter<ChatState> emit,
  ) async {
    // Add user message
    final userMessage = Message(
      id: DateTime.now().millisecondsSinceEpoch.toString(),
      kind: MessageKind.user,
      content: TextContent(event.message),
      timestamp: DateTime.now(),
    );

    final currentMessages = List<Message>.from(state.messages);
    currentMessages.add(userMessage);

    emit(state.copyWith(
      messages: currentMessages,
      status: ChatStatus.loading,
    ));

    try {
      final eventStream = _apiClient.sendMessage(
        conversationId: state.conversationId,
        message: event.message,
      );

      // Buffer for building up assistant message text
      String? currentAssistantMessageId;
      String assistantTextBuffer = '';
      int messageCounter = 0;

      await for (final clientEvent in eventStream) {
        _logger.d('Received client event: ${clientEvent.runtimeType}');

        switch (clientEvent) {
          case AssistantThinkingEvent():
            emit(state.copyWith(status: ChatStatus.assistantThinking));

          case AssistantMessageChunkEvent():
            // Build up assistant message incrementally
            assistantTextBuffer += clientEvent.chunk;

            // Create new message ID if we don't have one
            if (currentAssistantMessageId == null) {
              messageCounter++;
              currentAssistantMessageId =
                  'assistant_${DateTime.now().millisecondsSinceEpoch}_$messageCounter';

              // Create new message and append to list
              currentMessages.add(Message(
                id: currentAssistantMessageId,
                kind: MessageKind.assistant,
                content: TextContent(assistantTextBuffer),
                timestamp: DateTime.now(),
              ));
            } else {
              // Update the last message (which should be our current assistant message)
              final lastIndex = currentMessages.length - 1;
              if (lastIndex >= 0 &&
                  currentMessages[lastIndex].id == currentAssistantMessageId) {
                currentMessages[lastIndex] = Message(
                  id: currentAssistantMessageId,
                  kind: MessageKind.assistant,
                  content: TextContent(assistantTextBuffer),
                  timestamp: currentMessages[lastIndex].timestamp,
                );
              }
            }

            emit(state.copyWith(
              messages: List.from(currentMessages),
              status: ChatStatus.assistantTyping,
            ));

          case AssistantMessageCompleteEvent():
            // Clear buffer for next message
            currentAssistantMessageId = null;
            assistantTextBuffer = '';

            emit(state.copyWith(
              status: ChatStatus.idle,
            ));

          case ToolStartEvent():
            // Add tool call message
            final toolCallMessage = Message(
              id: clientEvent.toolId,
              kind: MessageKind.toolCall,
              content: ToolCallContent(
                toolName: clientEvent.toolName,
                input: clientEvent.input,
              ),
              timestamp: DateTime.now(),
            );
            currentMessages.add(toolCallMessage);

            emit(state.copyWith(
              messages: List.from(currentMessages),
              currentToolCall: clientEvent.toolName,
              status: ChatStatus.toolExecuting,
            ));

          case ToolResultEvent():
            // Add tool result message
            final toolResultMessage = Message(
              id: '${clientEvent.toolId}_result',
              kind: MessageKind.toolResult,
              content: ToolResultContent(
                toolName: clientEvent.toolName,
                result: clientEvent.result,
                error: clientEvent.error,
              ),
              timestamp: DateTime.now(),
            );
            currentMessages.add(toolResultMessage);

            emit(state.copyWith(
              messages: List.from(currentMessages),
              currentToolCall: null,
              status: ChatStatus.assistantThinking,
            ));

          case GeoFeatureEvent():
            // Add geo feature to map
            _mapBloc.add(AddGeoFeature(clientEvent.feature));

            // Optionally add as message too for chat history
            final geoMessage = Message(
              id: clientEvent.feature.id,
              kind: MessageKind.geoFeature,
              content: GeoFeatureContent(clientEvent.feature),
              timestamp: DateTime.now(),
            );
            currentMessages.add(geoMessage);

            emit(state.copyWith(messages: List.from(currentMessages)));

          case FeatureRemovedEvent():
            // Remove geo feature from map
            _mapBloc.add(RemoveGeoFeature(clientEvent.featureId));

          case ErrorEvent():
            throw Exception('${clientEvent.errorType}: ${clientEvent.message}');

          case DoneEvent():
            // Stream complete
            _logger.d('Stream complete');

          case UnknownEvent():
            _logger.w('Unknown event type: ${clientEvent.type}');
        }
      }

      emit(state.copyWith(
        messages: currentMessages,
        status: ChatStatus.idle,
        currentToolCall: null,
      ));
    } catch (e) {
      _logger.e('Error during message send: $e');

      final errorMessage = Message(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        kind: MessageKind.error,
        content: TextContent('Error: ${e.toString()}'),
        timestamp: DateTime.now(),
      );

      emit(state.copyWith(
        messages: [...currentMessages, errorMessage],
        status: ChatStatus.error,
        errorMessage: e.toString(),
        currentToolCall: null,
      ));
    }
  }

  @override
  Future<void> close() {
    // Don't dispose apiClient - it's shared and owned by the app
    return super.close();
  }
}

File: react_client/src/blocs/chat/chat_bloc.ts
// ABOUTME: BLoC managing chat state and API server communication.
// ABOUTME: Consumes agent event streams from API server for real-time updates.

import { Bloc } from "@granite-crisp/react-bloc";
import type { ChatEvent } from "./chat_event";
import { type ChatState, createInitialChatState } from "./chat_state";
import type { MapBloc } from "../map/map_bloc";
import { ApiClient } from "../../services/api_client";
import {
  type Message,
  createTextMessage,
  createToolCallMessage,
  createToolResultMessage,
  createSubagentCallMessage,
  createSubagentTextMessage,
  createSubagentResultMessage,
  createGeoFeatureMessage,
} from "../../models";

export class ChatBloc extends Bloc<ChatEvent, ChatState> {
  private readonly apiClient: ApiClient;
  private readonly mapBloc: MapBloc;

  constructor(apiClient: ApiClient, mapBloc: MapBloc) {
    super(createInitialChatState());
    this.apiClient = apiClient;
    this.mapBloc = mapBloc;
  }

  add(event: ChatEvent): void {
    switch (event.type) {
      case "setInputText":
        this.emit({ inputText: event.text });
        break;
      case "sendMessage":
        this.handleSendMessage();
        break;
    }
  }

  private async handleSendMessage(): Promise<void> {
    const messageText = this.state.inputText.trim();
    if (!messageText) return;

    // Add user message and clear input
    const userMessage = createTextMessage(
      String(Date.now()),
      "user",
      messageText
    );

    let messages: Message[] = [...this.state.messages, userMessage];

    this.emit({
      messages,
      inputText: "",
      status: "loading",
    });

    try {
      const eventStream = this.apiClient.sendMessage(
        this.state.conversationId,
        messageText
      );

      // Buffer for building up assistant message text
      let currentAssistantMessageId: string | null = null;
      let assistantTextBuffer = "";
      let messageCounter = 0;

      // Per-subagent streamed-text buffers, keyed by Task tool id
      // (subagents can run in parallel, so chunks interleave)
      const subagentTextBuffers = new Map<string, { id: string; text: string }>();

      for await (const event of eventStream) {
        switch (event.type) {
          case "assistant_thinking":
            this.emit({ status: "assistantThinking" });
            break;

          case "assistant_message_chunk":
            assistantTextBuffer += event.chunk;

            if (currentAssistantMessageId === null) {
              messageCounter++;
              currentAssistantMessageId = `assistant_${Date.now()}_${messageCounter}`;

              messages = [
                ...messages,
                createTextMessage(
                  currentAssistantMessageId,
                  "assistant",
                  assistantTextBuffer
                ),
              ];
            } else {
              // Update the last message with accumulated text
              const lastIndex = messages.length - 1;
              if (
                lastIndex >= 0 &&
                messages[lastIndex].id === currentAssistantMessageId
              ) {
                messages = [
                  ...messages.slice(0, lastIndex),
                  createTextMessage(
                    currentAssistantMessageId,
                    "assistant",
                    assistantTextBuffer
                  ),
                ];
              }
            }

            this.emit({
              messages,
              status: "assistantTyping",
            });
            break;

          case "assistant_message_complete":
            currentAssistantMessageId = null;
            assistantTextBuffer = "";
            this.emit({ status: "idle" });
            break;

          case "tool_start":
            messages = [
              ...messages,
              createToolCallMessage(event.toolId, event.toolName, event.input),
            ];
            this.emit({
              messages,
              currentToolCall: event.toolName,
              status: "toolExecuting",
            });
            break;

          case "tool_result":
            messages = [
              ...messages,
              createToolResultMessage(
                `${event.toolId}_result`,
                event.toolName,
                event.result,
                event.error
              ),
            ];
            this.emit({
              messages,
              currentToolCall: undefined,
              status: "assistantThinking",
            });
            break;

          case "subagent_started":
            messages = [
              ...messages,
              createSubagentCallMessage(event.toolId, event.agentName),
            ];
            this.emit({
              messages,
              currentSubagent: event.agentName,
              status: "subagentExecuting",
            });
            break;

          case "subagent_text_chunk": {
            let entry = subagentTextBuffers.get(event.toolId);
            if (!entry) {
              messageCounter++;
              entry = {
                id: `subagent_text_${event.toolId}_${messageCounter}`,
                text: event.chunk,
              };
              subagentTextBuffers.set(event.toolId, entry);
              messages = [
                ...messages,
                createSubagentTextMessage(
                  entry.id,
                  event.agentName,
                  entry.text
                ),
              ];
            } else {
              entry.text += event.chunk;
              const idx = messages.findIndex((m) => m.id === entry!.id);
              if (idx >= 0) {
                messages = [
                  ...messages.slice(0, idx),
                  createSubagentTextMessage(
                    entry.id,
                    event.agentName,
                    entry.text
                  ),
                  ...messages.slice(idx + 1),
                ];
              }
            }
            this.emit({ messages, status: "subagentExecuting" });
            break;
          }

          case "subagent_completed":
            messages = [
              ...messages,
              createSubagentResultMessage(
                `${event.toolId}_result`,
                event.agentName,
                event.error
              ),
            ];
            this.emit({
              messages,
              currentSubagent: undefined,
              status: "assistantThinking",
            });
            break;

          case "geo_feature":
            this.mapBloc.add({ type: "addGeoFeature", feature: event.feature });
            messages = [
              ...messages,
              createGeoFeatureMessage(event.feature.id, event.feature),
            ];
            this.emit({ messages });
            break;

          case "feature_removed":
            this.mapBloc.add({
              type: "removeGeoFeature",
              featureId: event.featureId,
            });
            break;

          case "error":
            throw new Error(`${event.errorType}: ${event.message}`);

          case "done":
            // Stream complete
            break;

          case "unknown":
            console.warn("Unknown event type:", event.eventType);
            break;
        }
      }

      this.emit({
        messages,
        status: "idle",
        currentToolCall: undefined,
        currentSubagent: undefined,
      });
    } catch (e) {
      console.error("Error during message send:", e);

      const errorMessage = createTextMessage(
        String(Date.now()),
        "error",
        `Error: ${e instanceof Error ? e.message : String(e)}`
      );

      this.emit({
        messages: [...messages, errorMessage],
        status: "error",
        errorMessage: e instanceof Error ? e.message : String(e),
        currentToolCall: undefined,
        currentSubagent: undefined,
      });
    }
  }
}
